chainerx_cc/chainerx/cuda/cuda_device/linalg.cu

```diff
@@ -0,0 +1,129 @@
+#include "chainerx/cuda/float16.cuh"
+#include "chainerx/cuda/kernel_regist.h"
+#include "chainerx/device.h"
+#include "chainerx/dtype.h"
+#include "chainerx/error.h"
+#include "chainerx/float16.h"
+#include "chainerx/kernels/creation.h"
+#include "chainerx/kernels/linalg.h"
+#include "chainerx/kernels/misc.h"
+#include "chainerx/macro.h"
+#include "chainerx/native/native_device.h"
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/linalg.h"
+
+namespace chainerx {
+namespace cuda {
+
+class CudaSolveKernel : public SolveKernel {
+public:
+    void Call(const Array& a, const Array& b, const Array& out) override {
+        Device& device = a.device();
+        Dtype dtype = a.dtype();
+        CudaSetDeviceScope scope{device.index()};
+
+        CHAINERX_ASSERT(a.ndim() == 2);
+        CHAINERX_ASSERT(a.shape()[0] == a.shape()[1]);
+
+        auto solve_impl = [&](auto pt, auto getrf_bufferSize, auto getrf, auto getrs) {
+            using T = typename decltype(pt)::type;
+            cuda_internal::DeviceInternals& device_internals = cuda_internal::GetDeviceInternals(static_cast<CudaDevice&>(device));
+
+            Array lu_matrix = Empty(a.shape(), dtype, device);
+            device.backend().CallKernel<CopyKernel>(a.Transpose(), lu_matrix);
+            T* lu_ptr = static_cast<T*>(internal::GetRawOffsetData(lu_matrix));
+
+            int m = a.shape()[0];
+
+            Array ipiv = Empty(Shape({m}), Dtype::kInt32, device);
+            int* ipiv_ptr = static_cast<int*>(internal::GetRawOffsetData(ipiv));
+
+            int buffersize = 0;
+            device_internals.cusolverdn_handle().Call(getrf_bufferSize, m, m, lu_ptr, m, &buffersize);
+
+            Array work = Empty(Shape({buffersize}), dtype, device);
+            T* work_ptr = static_cast<T*>(internal::GetRawOffsetData(work));
+
+            std::shared_ptr<void> devInfo = device.Allocate(sizeof(int));
+
+            device_internals.cusolverdn_handle().Call(getrf, m, m, lu_ptr, m, work_ptr, ipiv_ptr, static_cast<int*>(devInfo.get()));
+
+            int devInfo_h = 0;
+            Device& native_device = dynamic_cast<native::NativeDevice&>(GetDefaultContext().GetDevice({"native", 0}));
+            device.MemoryCopyTo(&devInfo_h, devInfo.get(), sizeof(int), native_device);
+            if (devInfo_h != 0) {
+                throw ChainerxError{"Unsuccessfull getrf (LU) execution. Info = ", devInfo_h};
+            }
+
+            device.backend().CallKernel<CopyKernel>(b, out);
+            T* out_ptr = static_cast<T*>(internal::GetRawOffsetData(out));
+
+            device_internals.cusolverdn_handle().Call(
+                    getrs, CUBLAS_OP_N, m, m, lu_ptr, m, ipiv_ptr, out_ptr, m, static_cast<int*>(devInfo.get()));
+
+            device.MemoryCopyTo(&devInfo_h, devInfo.get(), sizeof(int), native_device);
+            if (devInfo_h != 0) {
+                throw ChainerxError{"Unsuccessfull getrs (Solve) execution. Info = ", devInfo_h};
```

Comment:
When does this error occur?

Reply:
For[ `getrs`](https://software.intel.com/en-us/mkl-developer-reference-c-getri)
If `info = 0`, the execution is successful.
If `info = -i`, i-th parameter had an illegal value.
Probably this could be raised only in case when the implementation is really wrong.
`NaN` and `Inf` are not illegal values for LAPACK routines [(source)](https://software.intel.com/sites/products/documentation/doclib/mkl_sa/11/mkl_lapack_examples/lapackro2.htm)